Docs: Self-Hosted Satellites (#3791)
earthly/earthly: docs/cloud/satellites/self-hosted.md

# Self-Hosted Satellites

> Self-hosted satellites _\*Beta\*_ are a good alternative to Earthly Cloud’s [managed satellites](../satellites.md) when CI/CD is required to run in your own environment. Self-hosted satellites provide most of the benefits of Earthly Cloud while ensuring that build-related data never leaves your network.

Reviewer: ```suggestion
Self-hosted satellites **Beta** are a good alternative to Earthly Cloud’s [managed satellites](../satellites.md) when CI/CD is required to run in your own environment. Self-hosted satellites provide most of the benefits of Earthly Cloud while ensuring that your code and build-related data never leaves your network.
```

In addition, Self-Hosted Satellites have the following features:
* Lightning-fast caching
* Cloud-enabled control plane
* Encrypted connections by default using mTLS

> * Ready to run builds with just a single command

Reviewer: Perhaps we should mention only the benefits specific to self-hosted satellites, otherwise we're repeating content from https://docs.earthly.dev/earthly-cloud/satellites.

Looping in @thtmnisamnstr in case he has thoughts on messaging here.

Reviewer: Side note: We should also amend step 2 from the page I linked above to say something along the lines of "if you want to launch a self-hosted satellite, go to this other page". Otherwise all the steps on that page are the same.

Reviewer 2: Replying to this comment: "Perhaps we should mention only the benefits specific to self-hosted satellites, otherwise we're repeating content from https://docs.earthly.dev/earthly-cloud/satellites."

I removed the "In addition to, ..." wording and updated copy a little, and I like it better. I don't think reiterating the benefits here is a bad thing.

> On the other hand, they may have the following drawbacks when compared to managed satellites offered in Earthly Cloud:

Reviewer: @thtmnisamnstr - thoughts on what to call the regular satellites? I think we use the term "Cloud" on our pricing. Do we then call these "Cloud Satellites" instead of managed satellites?

Reviewer 2: That makes sense. Cloud Satellites, Self-Hosted Satellites, and Dedicated Satellites.

* No automatic updates
* Does not automatically sleep to save costs while idle
* Requires more knowledge and tuning to achieve good performance

## Getting Started

> Launching a self-hosted satellite is as easy as running the public Docker image. A few environment variables are needed to link the satellite with your Earthly account. The satellite will use these values to automatically register with Earthly Cloud servers once it starts. Note that Earthly Cloud will never connect to your satellite directly.

Reviewer: ```suggestion
Launching a self-hosted satellite is as easy as running the public Docker image. A few environment variables are needed to link the satellite with your Earthly account. The satellite will use these values to automatically register with Earthly Cloud servers once it starts. Note that Earthly Cloud will never connect to your satellite directly. The host and port of your satellite can thus be private to your internal corporate network or VPC, if you wish.
```

Here is a minimal command to start a self-managed satellite using Docker:

```
docker run --privileged \
    -p 8372:8372 \
    -e EARTHLY_TOKEN=GuFna*****nve7e \ 
    -e EARTHLY_ORG=my-org \
    -e SATELLITE_NAME=my-satellite \
    -e SATELLITE_HOST=153.65.8.0 \

>   earthly/satellite:v0.8.3

Reviewer: Can we get this line to auto-update via renovate?

```

The following environment variables are required:
* `EARTHLY_TOKEN` - a persistent auth token obtained by running `earthly account create-token <token-name>`
* `EARTHLY_ORG` - the name of your Earthly Organization created on [https://cloud.earthly.dev](https://cloud.earthly.dev)
* `SATELLITE_NAME` - a name chosen by the user to identify the satellite instance
* `SATELLITE_HOST` - Hostname or IP address for the Earthly CLI to connect. Note that users will select the satellite by name when running builds.

Using a volume for cache storage is recommended for preserving cache after the container is destroyed. See [Advanced Configuration](#advanced-configuration) below for more details and a list of optional environment variables that can be used to finetune your deployment. See also [Platform Specific Guides](#platform-specific-guides) for examples of deploying to platforms like Kubernetes or EC2.

The self-hosted satellite will not accept localhost as an address. For instructions on testing locally, please refer to [Running on Localhost](#running-on-localhost) below.

{% hint style='info' %}
##### Security Advisory

> The `--privileged` flag is currently required. Privileged mode is used by some features of earthly, including the `WITH DOCKER` command in Earthfiles. If using privileged mode is a concern, running satellites in a dedicated VM or Kubernetes cluster can provide better isolation compared to running the container directly in your production environment.

Reviewer: ```suggestion
The `--privileged` flag is currently required. Privileged mode is used by some features of earthly, including the `WITH DOCKER` command in Earthfiles. If using privileged mode is a concern, running satellites in a dedicated VM or separate Kubernetes cluster can provide better isolation compared to running the container directly in your production environment.
```

Reviewer: I would perhaps add that a rootless version will be available in the future.

{% endhint %}

### Auto-discovery of Host _\*experimental\*_

> When `SATELLITE_HOST` is left unset, the satellite will attempt to auto-discover its host address. This feature may not always work, and currently only supports environments using cloud-init (such as EC2).

Reviewer: ```suggestion
When `SATELLITE_HOST` is left unset, the satellite will attempt to auto-discover its host address. Auto-discovery may not always work, and currently only supports environments using cloud-init (such as EC2).
```

## Connecting to a Self-Hosted Satellite

Once the satellite has finished registering, it can be selected and managed using typical earthly satellite commands from the CLI. 
For example, to select the satellite for use in subsequent builds:

```
earthly sat select my-satellite
```

Or, to invoke a build once on the satellite, without persistently selecting it:

```
earthly --sat my-satellite +my-target
```

> Note that `earthly-v0.8.0` or later is required to connect to a self-hosted satellite.

Reviewer: ```suggestion
Note that `earthly` `v0.8.0` or later is required to connect to a self-hosted satellite.
```

## Managing Self-Hosted Satellites

A list of satellites in your org be viewed via the satellite ls command. For example:

```
> earthly sat ls
NAME          PLATFORM     SIZE         VERSION  STATE
my-satellite  linux/amd64  self-hosted  v0.8.3   operational
...
```

The size field of a self-hosted satellite will be listed as `self-hosted`, as opposed to one of the sizes of Earthly Cloud’s managed satellites.

The state field will be either `operational` or `offline`. A self-hosted satellite automatically transitions to an `offline` state when it is gracefully terminated, or if no heartbeat has been received by Earthly’s servers for a while.

> You can run `satellite rm` on a self-hosted satellite when it is in an offline state to remove it from your account permenantly.

Reviewer: Add mention of `--force`?

Author: Not sure if we should promote that for self-hosted. I think of it as an escape hatch just in case something gets stuck in a bad state, as opposed to normal operation.

Reviewer: It's probably a very useful escape hatch - sometimes you don't have control over your infrastructure and you really need to disconnect it. Anyway... it's a small feature, we can also just leave it out and let them find it via `rm` attempt - presumably the error says that you can use `--force`.

Author: Yes, it will say that they can re-run with force if they try it. 

Also (in case it wasn't obvious) this wouldn't trigger the actual self-hosted satellite to terminate. Just remove it from the list on earthly side.

## Platform-Specific Guides

### Kubernetes

Below is a basic example of how to start a self-hosted satellite as a Kubernetes Pod:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-satellite
  namespace: earthly-satellites
spec:
  containers:
   - name:
   valueFrom:
     fieldRef:
       fieldPath: metadata.name
      image: earthly/satellite:v0.8.3
      securityContext:
        privileged: true
      ports:
        - containerPort: 8372
       env:
        - name: EARTHLY_ORG
          value: my-org
        - name: EARTHLY_TOKEN
          value: u4a*************l92
        - name: SATELLITE_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: SATELLITE_HOST
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
```

Note that this example uses the Pod’s IP address (via [Downward API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/)) as the `SATELLITE_HOST` value. The Earthly CLI must be able to reach the IP on its network.

If the Earthly CLI should connect via a different address, such as DNS, then this value should be provided as the `SATELLITE_HOST` instead.

> ### AWS EC2

Reviewer: I wonder if the EC2 version should be more encouraged in this guide since VMs provide a better security sandbox for builds. Just a thought.

Author: People seem very inclined to use K8S, but what if we put a "recommended" label on AWS , and/or bump it higher in the doc?

Reviewer: SGTM

Reviewer 2: That makes sense to me too. EC2 should be above k8s and encouraged.

When launching on EC2, we recommend using the latest version of Amazon Linux 2023. The following cloud-init script can be configured when launching a new EC2 instance so that it automatically starts the satellite on boot.

```yaml
#cloud-config
runcmd:
  - sudo dnf update
  - sudo dnf install -y docker
  - sudo systemctl start docker.service
  - sudo systemctl enable docker.service
  - |-
      sudo docker run -d --privileged \
        --restart always \
        --name satellite \
        -p 8372:8372 \
        -v /earthly-cache:/tmp/earthly:rw \
        -e EARTHLY_TOKEN=GuFna*****nve7e \
        -e EARTHLY_ORG=my-org \
        -e SATELLITE_NAME=my-satellite \
        earthly/satellite:v0.8.3
```

Note that the `SATELLITE_HOST` variable is unset in this example so that the host is auto-discovered by the satellite when it starts. This should result in the instance’s private DNS being used as the host. 

If the Earthly CLI will be unable to connect to the satellite via the EC2’s private DNS, then `SATELLITE_HOST` should be provided in the `docker run` command with an alternate value.

## Advanced Configuration

> ### Cache Disk

Reviewer: I think that our position should be stronger here. Not using a volume causes extremely bad `WITH DOCKER` performance (due to overlayfs). We should warn users that they MUST use a volume for this, and that volume should be backed by an ext4 filesystem (Docker volumes work great).

All our examples above should include the volume mount (the minimal docker run command and k8s example).

Author: I didn't realize there were legitimate issues with not using a volume. I'll put this in the other examples as well then.

* `RUNNER_DISABLE_TLS` - Disable TLS on the satellite. Requires Earthly CLI to also disable TLS. Not recommended in most cases.
* `LOG_LEVEL` - The log level of the internal "runner" process within the satellite. Set to INFO by default.
* `BUILDKIT_DEBUG` - Enable buildkit debug-level logs. False by default.
* `EARTHLY_ADDITIONAL_BUILDKIT_CONFIG` - allows additional buildkit configs to be injected in TOML format.

## Running on Localhost

For testing purposes, you may want to try your self-hosted satellite on localhost. Satellites currently do not support using `localhost` or `127.0.0.1` as an address when supplied to `SATELLITE_HOST`, in part because Earthly CLI has reserved this address for use as its own local Buildkit container.

It’s still possible to test self-hosted satellites locally, however, by using an alternate entry in your `/etc/hosts` file that maps to `localhost`. For example, you can try adding this entry:

```
# Local test for Earthly
127.0.0.1	earthly.local
```

Starting your satellite with `SATELLITE_HOST` set to `earthly.local` should allow for a localhost connection using your Earthly CLI.

## Debugging

> If you are having problems using or deploying your self-hosted satellite, please refer to the following tips or reach out to us through our community Slack channel.

Reviewer: ```suggestion
If you are having problems using or deploying your self-hosted satellite, please refer to the following tips or reach out to us through our community [Slack channel](https://earthly.dev/slack).
```

> **Problem:** Satellite is not listed in the output of earthly satellite ls

Reviewer: ```suggestion
### Problem: Satellite is not listed in the output of `earthly satellite ls`
```

Ditto for all the others...

> **Resolution:** Check the logs from the satellite’s Docker container. There may be a message containing the phrase "SATELLITE IS NOT REGISTERED". This usually means there was a problem with the values supplied to the satellite’s run command. Check the error for additional context. Ensure the values provided for account token, earthly org, etc are correct.

Reviewer: ```suggestion
**Resolution:** Check the logs from the satellite’s Docker container. There may be a message containing the phrase `SATELLITE IS NOT REGISTERED`. This usually means there was a problem with the values supplied to the satellite’s run command. Check the error for additional context. Ensure the values provided for account token, earthly org, etc are correct.
```
